### PR TITLE
[WIP] TAP Win32 rewrite

### PR DESCRIPTION
**Pull request by zeroZshadow**
Now closely resembles the unix implementation
Code can be merged into a shared base later

### CODE REVIEW

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 1]
```diff
@@ -242,19 +263,19 @@ bool CEXIETHERNET::IsActivated()
 
 bool CEXIETHERNET::SendFrame(u8 *frame, u32 size)
```
Comment (booto) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Any reason why `frame` can't be `const u8*`?

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 2]
```diff
@@ -242,19 +263,19 @@ bool CEXIETHERNET::IsActivated()
...
-	ZeroMemory(&overlap, sizeof(overlap));
+	// Make sure our first write is done before we start the second
+	WaitForSingleObject(mHWriteEvent, INFINITE);
 
 	// WriteFile will always return false because the TAP handle is async
-	WriteFile(mHAdapter, frame, size, NULL, &overlap);
-
-	DWORD res = GetLastError();
-	if (res != ERROR_IO_PENDING)
+	if (!WriteFile(mHAdapter, frame, size, NULL, &mWriteOverlapped))
```
Comment (booto) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: From the [docs on WriteFile](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365747%28v=vs.85%29.aspx):
> Accessing the output buffer while a write operation is using the buffer may lead to corruption of the data written from that buffer. Applications must not write to, reallocate, or free the output buffer that a write operation is using until the write operation completes.
I'm not sure about the semantics of the caller, but is seems like this code doesn't do anything to guarantee this?

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 3]
```diff
@@ -263,78 +284,69 @@ bool CEXIETHERNET::SendFrame(u8 *frame, u32 size)
...
 {
-	CEXIETHERNET* self = (CEXIETHERNET*)lpParameter;
+	while (true)
+	{
+		if (self->mHAdapter == INVALID_HANDLE_VALUE)
+			return;
 
-	GetOverlappedResult(self->mHAdapter, &self->mReadOverlapped,
-		(LPDWORD)&self->mRecvBufferLength, false);
+		DWORD res = ReadFile(self->mHAdapter, self->mRecvBuffer, BBA_RECV_SIZE, (LPDWORD)&self->mRecvBufferLength, &self->mReadOverlapped);
```
Comment (booto) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Is the cast to `LPDWORD` (I believe that's the equivalent of `unsigned int*`) necessary here?

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 4]
```diff
@@ -263,78 +284,69 @@ bool CEXIETHERNET::SendFrame(u8 *frame, u32 size)
...
+				// Unexpected error
+				ERROR_LOG(SP1, "Failed to recieve packet with error 0x%X", err);
+				return;
+			}
+
+			// Wait for writing to be completed
+			//if (WaitForSingleObject(self->mHRecvEvent, 50) != WAIT_OBJECT_0)
+			//	continue;
+
+			GetOverlappedResult(self->mHAdapter, &self->mReadOverlapped, (LPDWORD)&self->mRecvBufferLength, true);
```
Comment (booto) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Same as above (is cast necessary?)

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 5]
```diff
@@ -242,99 +271,87 @@ bool CEXIETHERNET::IsActivated()
...
+		if (err != ERROR_IO_PENDING)
+		{
+			ERROR_LOG(SP1, "Failed to send packet with error 0x%X", err);
+		}
 	}
 
 	// Always report the packet as being sent successfully, even though it might be a lie
 	SendComplete();
-
+	INFO_LOG(SP1, "SendFrame complete");
```
Comment (shuffle2) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Logging is slow, so try not to introduce logs which are present in release builds in the send/recv paths. (iirc INFO_LOG is active in release, might be wrong)

Comment (zeroZshadow) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Info doesnt show up. Also this is temp stuff for profiling.

**Comment by shuffle2** [comment]
Uh, you replaced Windows overlapped events with a busy waiting thread? Overall this is a -1 from me...

**Comment by zeroZshadow** [comment]
Yes, because this is exactly how the unix version does it.  Also the read will have it block, so its not consuming all cpu.

File: Source/Core/Core/HW/EXI_DeviceEthernet.h [hunk 6]
```diff
@@ -323,14 +323,12 @@ class CEXIETHERNET : public IEXIDevice
 	u32 mRecvBufferLength;
 
 #if defined(_WIN32)
-	HANDLE mHAdapter, mHRecvEvent, mHReadWait;
-	DWORD mMtu;
-	OVERLAPPED mReadOverlapped;
-	static VOID CALLBACK ReadWaitCallback(PVOID lpParameter, BOOLEAN TimerFired);
+	HANDLE mHAdapter, mHReadEvent, mHWriteEvent;
```
Comment (shuffle2) [inline Source/Core/Core/HW/EXI_DeviceEthernet.h]: mHReadEvent and mReadOverlapped are now unused.

**Comment by mathieui** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 7]
```diff
@@ -221,6 +222,22 @@ bool CEXIETHERNET::Activate()
...
+	{
+		ERROR_LOG(SP1, "Failed to create write event:%x", GetLastError());
+		return false;
+	}
+
+	// Initialize overlapped
+	ZeroMemory(&mWriteOverlapped, sizeof(mWriteOverlapped));
+	mWriteOverlapped.hEvent = mHWriteEvent;
+
+	// Start early so we can get all those pestky
```
Comment (mathieui) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Trailing WS

File: Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp [hunk 8]
```diff
@@ -242,99 +269,108 @@ bool CEXIETHERNET::IsActivated()
...
+		DWORD err = GetLastError();
+		if (err && err != ERROR_IO_PENDING)
+		{
+			ERROR_LOG(SP1, "Failed to send packet with error 0x%X", err);
+		}
+
+		DWORD byteswritten;
+		GetOverlappedResult(mHAdapter, &mWriteOverlapped, &byteswritten, true);
 	}
+
```
Comment (mathieui) [inline Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp]: Trailing WS

**Comment by mathieui** [comment]
@dolphin-emu-bot rebuild

**Comment by mathieui** [comment]
@dolphin-emu-bot rebuild

**Comment by mathieui** [comment]
@dolphin-emu-bot rebuild

**Comment by zeroZshadow** [comment]
No longer going to bother with window's horrible wait of reading and writing devices. If anyone wants to give this a try, give me a poke and ill share my gained knowledge thus far.

**Pull request closed without merging**